source/slang/legalize-types.h

```diff
@@ -328,20 +328,139 @@ RefPtr<TypeLayout> getDerefTypeLayout(
+    ///
+    /// Operations like computing an offset can then be encoded by
+    /// starting with zero and then walking up the chain and adding in
+    /// offsets as encountered.
+    ///
+struct SimpleLegalVarChain
+{
+    // The next link up the chain, or null if this is the end.
+    SimpleLegalVarChain*    next = nullptr;
+
+    // The layout for the variable at this link in thain.
+    VarLayout*              varLayout = nullptr;
+};
+
+    /// A "chain" of variable declarations that can handle both primary and "pending" data.
+    ///
+    /// In the presence of interface-type fields, a single variable may
+    /// have data that sits in two distinct allocations, and may have
+    /// `VarLayout`s that represent offseting into each of those
+    /// allocations.
+    ///
+    /// A `LegalVarChain` tracks two distinct `SimpleVarChain`s: one for
+    /// the primary/ordinary data allocation, and one for any pending
+    /// data.
+    ///
+    /// It is okay if the primary/pending chains have different numbers
+    /// of links in them.
+    ///
+    /// Offsets for particular resource kinds in the primary or pending
+    /// data allocation can be queried on the appropriate sub-chain.
+    ///
 struct LegalVarChain
 {
-    LegalVarChain*  next;
-    VarLayout*      varLayout;
+    // The chain of variables that represents the primary allocation.
+    SimpleLegalVarChain*    primaryChain = nullptr;
+
+    // The chain of variables that represents the pending allocation.
+    SimpleLegalVarChain*    pendingChain = nullptr;
+
+    // If the primary chain is non-empty, gets the variable at the leaf.
+    DeclRef<VarDeclBase> getLeafVarDeclRef() const
+    {
+        if(!primaryChain)
+            return DeclRef<VarDeclBase>();
+
+        return primaryChain->varLayout->varDecl;
+    }
+};
+
+    /// RAII type for adding a link to a `LegalVarChain` as needed.
+    ///
+    /// This type handles the bookkeeping for creating a `LegalVarChain`
+    /// that links in one more variable. It will add a link to each of
+    /// the primary and pending sub-chains if and only if there is non-null
+    /// layout information for the primary/pending case.
+    ///
+    /// Typical usage in a recursive function is:
+    ///
+    ///     void someRecursiveFunc(LegalVarChain const& outerChain, ...)
+    ///     {
+    ///         if(auto subVar = needToRecurse(...))
+    ///         {
+    ///             LegalVarChainLink subChain(outerChain, subVar);
+    ///             someRecursiveFunc(subChain, ...);
+    ///         }
+    ///         ...
+    ///     }
+    ///
+struct LegalVarChainLink : LegalVarChain
+{
+        /// Default constructor: yields an empty chain.
+    LegalVarChainLink()
+    {
+    }
+
+        /// Copy constructor: yields a copy of the `parent` chain.
+    LegalVarChainLink(LegalVarChain const& parent)
+        : LegalVarChain(parent)
+    {}
+
+        /// Construct a chain that extends `parent` with `varLayout`, if it is non-null.
```

Comment:
That is copy constructs its 'parent' into its base class seems unexpected.

Reply:
It makes sense that if we are conditionally adding a new link to a linked list, then the default value (in case we *don't* end up adding anything) is the existing linked list, which would also be the "parent" pointer / next pointer if we *do* add something.

I'd welcome better terminology here, but this ends up being a very specific data structure for allowing the algorithms to be written conveniently, and not a major piece of the public API.